app/Image/SizeVariantDefaultFactory.php

```diff
@@ -0,0 +1,361 @@
+
+			return null;
+		}
+	}
+
+	protected function createSizeVariantInternal(int $sizeVariant, int $maxWidth, int $maxHeight): SizeVariant
+	{
+		$shortPath = $this->namingStrategy->generateShortPath($sizeVariant);
+
+		$sv = $this->photo->size_variants->getSizeVariant($sizeVariant);
+		if (!$sv) {
+			$sv = $this->photo->size_variants->createSizeVariant($sizeVariant, $shortPath, $maxWidth, $maxHeight);
+			if ($sizeVariant === SizeVariant::THUMB || $sizeVariant === SizeVariant::THUMB2X) {
+				$success = $this->imageHandler->crop($this->referenceFullPath, $sv->full_path, $sv->width, $sv->height);
+			} else {
+				$resWidth = $resHeight = 0;
+				$success = $this->imageHandler->scale($this->referenceFullPath, $sv->full_path, $sv->width, $sv->height, $resWidth, $resHeight);
+				$sv->width = $resWidth;
+				$sv->height = $resHeight;
+				$sv->save();
+			}
+			if (!$success) {
+				Logs::error(__METHOD__, __LINE__, 'Failed to resize image: ' . $this->referenceFullPath);
+				// If scaling/cropping has failed, remove the freshly create DB entity again
+				// This will also take care of removing a potentially created file from storage
+				$sv->delete();
+				throw new \RuntimeException('Failed to resize image: ' . $this->referenceFullPath);
+			}
+		}
+
+		return $sv;
+	}
+
+	/**
+	 * Determines the maximum dimensions of the designated size variant.
+	 *
+	 * @param int $sizeVariant the size variant
+	 *
+	 * @return int[] an array with exactly two integers, the first integer is
+	 *               the width, the second integer is the height
+	 */
+	protected function getMaxDimensions(int $sizeVariant): array
+	{
+		switch ($sizeVariant) {
+			case SizeVariant::MEDIUM2X:
+				$maxWidth = 2 * intval(Configs::get_value('medium_max_width'));
+				$maxHeight = 2 * intval(Configs::get_value('medium_max_height'));
+				break;
+			case SizeVariant::MEDIUM:
+				$maxWidth = intval(Configs::get_value('medium_max_width'));
+				$maxHeight = intval(Configs::get_value('medium_max_height'));
+				break;
+			case SizeVariant::SMALL2X:
+				$maxWidth = 2 * intval(Configs::get_value('small_max_width'));
+				$maxHeight = 2 * intval(Configs::get_value('small_max_height'));
+				break;
+			case SizeVariant::SMALL:
+				$maxWidth = intval(Configs::get_value('small_max_width'));
+				$maxHeight = intval(Configs::get_value('small_max_height'));
+				break;
+			case SizeVariant::THUMB2X:
+				$maxWidth = self::THUMBNAIL2X_DIM;
+				$maxHeight = self::THUMBNAIL2X_DIM;
+				break;
+			case SizeVariant::THUMB:
+				$maxWidth = self::THUMBNAIL_DIM;
+				$maxHeight = self::THUMBNAIL_DIM;
+				break;
+			default:
+				throw new \InvalidArgumentException('unknown size variant: ' . $sizeVariant);
+		}
+
+		return [$maxWidth, $maxHeight];
+	}
+
+	protected function isEnabledByConfiguration(int $sizeVariant): bool
+	{
+		switch ($sizeVariant) {
+			case SizeVariant::MEDIUM2X:
+				return Configs::get_value('medium_2x', 0) == 1;
```

Comment:
The config values are string. Unless there is a cast that I am not aware off.

Reply:
I just inherited the code and due to PHP type coercion this actually works (note it is `==` not `===`). So `('1' == 1) === true` holds.

But indeed, I believe the whole `Configs` should be refactored to and have proper getter methods for each configuration option. So in this case it would be `Configs::isMedium2xEnabled(): boolean`. This would allow proper auto-completion by the IDE, it would make the code more robust, because PHP would throw an exception if the method name was misspelled and it would  also remove all this magic constants like `'medium_2x'` from the code. (The string would become proper constants of the `Configs` class).

Anyway this is another PR and needs re-factoring. But yes, I actually thought about that, but this PR is too huge as it is.

Reply:
> proper getter methods for each configuration option. So in this case it would be `Configs::isMedium2xEnabled(): boolean`

Having a method for every config option sounds very heavy. Especially as iirc we store types in the config table. If you wanted better type checking you could always call them as e.g. `getboolean` which would error if the setting wasn't a boolean?